include/symbolizer.hpp

```diff
@@ -0,0 +1,131 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#pragma once
+
+#include "datadog/blazesym.h"
+#include "ddprof_defs.hpp"
+#include "ddprof_file_info-i.hpp"
+#include "ddres_def.hpp"
+#include "map_utils.hpp"
+#include "mapinfo_table.hpp"
+
+#include <memory>
+#include <span>
+#include <string>
+#include <string_view>
+#include <vector>
+
+struct ddog_prof_Location;
+
+namespace ddprof {
+class Symbolizer {
+public:
+  enum AddrFormat {
+    k_elf,
+    k_process,
+  };
+
+  explicit Symbolizer(bool inlined_functions = false,
+                      bool disable_symbolization = false,
+                      AddrFormat reported_addr_format = k_process)
+      : inlined_functions(inlined_functions),
+        _disable_symbolization(disable_symbolization),
+        _reported_addr_format(reported_addr_format) {}
+
+  struct BlazeResultsWrapper {
+    BlazeResultsWrapper() = default;
+    ~BlazeResultsWrapper() {
+      for (auto &result : blaze_results) {
+        blaze_result_free(result);
+      }
+    }
+
+    // Delete copy constructor and copy assignment operator
+    BlazeResultsWrapper(const BlazeResultsWrapper &) = delete;
+    BlazeResultsWrapper &operator=(const BlazeResultsWrapper &) = delete;
+
+    // Optionally define move constructor and move assignment operator
+    BlazeResultsWrapper(BlazeResultsWrapper &&other) noexcept
+        : blaze_results(std::move(other.blaze_results)) {
+      other.blaze_results.clear();
+    }
+
+    BlazeResultsWrapper &operator=(BlazeResultsWrapper &&other) noexcept {
+      if (this != &other) {
+        for (auto &result : blaze_results) {
+          blaze_result_free(result);
+        }
+        blaze_results = std::move(other.blaze_results);
+        other.blaze_results.clear();
+      }
+      return *this;
+    }
+
+    std::vector<const blaze_result *> blaze_results{};
+  };
+
+  /// Fills the locations at the write index using address and elf source.
+  /// assumption is that all addresses are from this source file
+  /// Parameters
+  /// addrs - Elf address
+  /// process_addrs - Process address (only used for pprof reporting)
+  /// file_id - a way to identify this file in a unique way
+  /// elf_src - a path to the source file (idealy stable)
+  /// map_info - the mapping information to write to the pprof
+  /// locations - the output pprof strucure
+  /// write_index - input / output parameter updated based on what is written
+  /// results - A handle object for lifetime of strings.
+  ///          Should be kept until interned strings are no longer needed.
+  DDRes symbolize_pprof(std::span<ElfAddress_t> addrs,
+                        std::span<ProcessAddress_t> process_addrs,
+                        FileInfoId_t file_id, const std::string &elf_src,
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏ it looks like `elf_src` has a `/proc/<pid>/root` prefix. Does this prevent blazesym from using its cache across pids ?

Reply:
We have a wrapper that ensures we always access the file with the same path (including the /proc/<pid>). 
This should ensure we are able to mutualize the caches. Otherwise this was indeed an issue.
Ideally an upstream contribution would allow us to have better control over the caching logic. Maybe by providing the file descriptor or inode as key ?